### PR TITLE
Remove the `Gateway` interface

### PR DESCRIPTION
**Pull request by frouioui**
## Description

In https://github.com/vitessio/vitess/pull/9500 we removed the legacy gateway (`DiscoveryGateway`), leaving `TabletGateway` as the only implementation of the `vtgate/Gateway` interface. The interface is now useless, this pull request removes it, along with the `-gateway_implementation` flag that was previously used to decide which implementation of the gateway we wanted to use.

This pull request is rebased on top of #9500, do not merge until #9500 is merged. Commits d0941b1cd14397dd5df4ea323f208bd58e154cbf and 60470e7df9e3537af1e7c58ceaaf2a8a1442e710 are the last commits to look at.

## Related Issue(s)
- Part of #9218

## Checklist
- [ ] Should this PR be backported? _No_
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/tabletgateway.go [hunk 1]
```diff
@@ -177,7 +201,8 @@ func (gw *TabletGateway) WaitForTablets(ctx context.Context, tabletTypesToWait [
 	if err != nil {
 		return err
 	}
-	return gw.hc.WaitForAllServingTablets(ctx, targets)
+	err = gw.hc.WaitForAllServingTablets(ctx, targets)
+	return err
```
Comment (deepthi) [inline go/vt/vtgate/tabletgateway.go:205]: nit: this change seems unnecessary

Comment (frouioui) [reply go/vt/vtgate/tabletgateway.go:203]: correct

File: doc/TabletRouting.md [hunk 2]
```diff
@@ -78,28 +78,25 @@ This module also provides helper methods to find a tablet to route traffic
 to. Since it knows the health status of all tablets for a given keyspace / shard
 / tablet type, and their replication lag, it can provide a good list of tablets.
 
-## Vtgate Gateway interface
+## Vtgate TabletGateway
 
-An important interface inside vtgate is the Gateway interface. It can send
+An important interface inside vtgate is the TabletGateway. It can send
```
Comment (deepthi) [inline doc/TabletRouting.md:83]: not an interface any more, right?
maybe we should call it `component`.

File: doc/TabletRouting.md [hunk 3]
```diff
@@ -78,28 +78,25 @@ This module also provides helper methods to find a tablet to route traffic
...
+to route them to the right tablet.
 
-There are two implementations of the Gateway interface:
-
-* discoveryGateway: deprecated
-* tabletGateway: Combines a set of TopologyWatchers (described in the
-  discovery section, one per cell) as a source of tablets, a HealthCheck module
-  to watch their health, and a tabletHealthCheck per tablet to collect all the health
-  information. Based on this data, it can find the best tablet to use.
+The TabletGateway Combines a set of TopologyWatchers (described in the
```
Comment (deepthi) [inline doc/TabletRouting.md:90]:
```suggestion
TabletGateway combines a set of TopologyWatchers (described in the
```

File: doc/TabletRouting.md [hunk 4]
```diff
@@ -78,28 +78,25 @@ This module also provides helper methods to find a tablet to route traffic
...
+discovery section, one per cell) as a source of tablets, a HealthCheck module
+to watch their health, and a tabletHealthCheck per tablet to collect all the health
+information. Based on this data, it can find the best tablet to use.
   
 # Extensions, work in progress
 
 ## Regions, cross-cell targeting
 
-At first, the discoveryGateway was routing queries the following way:
+At first, the gateway was routing queries the following way:
```
Comment (deepthi) [inline doc/TabletRouting.md:99]: We need to rewrite this section. We don't have regions or l2vtgate any more.
Instead we have cell-aliases that group cells together.

File: go/vt/vtgate/tabletgateway.go [hunk 5]
```diff
@@ -45,6 +45,11 @@ var (
 	_ discovery.HealthCheck = (*discovery.HealthCheckImpl)(nil)
 	// CellsToWatch is the list of cells the healthcheck operates over. If it is empty, only the local cell is watched
 	CellsToWatch = flag.String("cells_to_watch", "", "comma-separated list of cells for watching tablets")
+
+	bufferImplementation = flag.String("buffer_implementation", "keyspace_events", "Allowed values: healthcheck (legacy implementation), keyspace_events (default)")
+	initialTabletTimeout = flag.Duration("gateway_initial_tablet_timeout", 30*time.Second, "At startup, the tabletGateway will wait up to that duration to get one tablet per keyspace/shard/tablettype")
```
Comment (deepthi) [inline go/vt/vtgate/tabletgateway.go:50]:
```suggestion
	initialTabletTimeout = flag.Duration("gateway_initial_tablet_timeout", 30*time.Second, "At startup, the tabletGateway will wait up to this duration to get at least one tablet per keyspace/shard/tablet type")
```

File: go/test/endtoend/cellalias/cell_alias_test.go [hunk 6]
```diff
@@ -226,72 +226,6 @@ func TestMain(m *testing.M) {
...
-	err = localCluster.VtctlclientProcess.ExecuteCommand("UpdateCellsAlias", "--",
-		"--cells", allCells,
-		"region_east_coast")
-	require.NoError(t, err)
-
-	vtgateInstance := localCluster.NewVtgateInstance()
-	vtgateInstance.CellsToWatch = allCells
-	vtgateInstance.TabletTypesToWait = "PRIMARY,REPLICA"
-	// Use legacy gateway. There's a separate test for tabletgateway in go/test/endtoend/tabletgateway/cellalias/cell_alias_test.go
-	vtgateInstance.GatewayImplementation = "discoverygateway"
```
[the remark this answers is not in the document]
Comment (frouioui) [reply go/test/endtoend/cellalias/cell_alias_test.go:262]: good point!

File: doc/TabletRouting.md (hunk 4 above)
Comment (deepthi) [reply doc/TabletRouting.md:99]: The entire paragraph on l2vtgate can be deleted.
The section on Region should be either deleted, or rewritten to talk about cell-aliases.
I'm fine with either approach.
We will need to change the heading that says "Regions, ..."

Comment (deepthi) [reply doc/TabletRouting.md:99]: Ref: https://github.com/vitessio/vitess/pull/4767

Comment (frouioui) [reply doc/TabletRouting.md]: I removed both sections via 0785e4a

**Review by deepthi — APPROVED** [review]

**Pull request merged**